jina/drivers/multimodal.py

```diff
@@ -0,0 +1,94 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from typing import Iterable, Dict, Tuple
+
+import numpy as np
+
+from .reduce import ReduceDriver
+from .helper import pb2array, array2pb
+from ..proto import jina_pb2
+
+class MultimodalDriver(ReduceDriver):
+    """
+    TODO add docstring
+    each document have multiple chunks
+    each chunk has 1 modality
+    group chunk i
+    """
+    def __init__(self, traversal_paths: Tuple[str] = ('c', ), *args, **kwargs):
+        # traversal chunks from chunk level.
+        # TODO discuss should we add root path
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
+
+    def reduce(self, *args, **kwargs) -> None:
+        doc_pointers = {}
+        # traverse apply on ALL requests collected to collect embeddings
+        # reversed since the last response should collect the chunks/matches
+        for r in reversed(self.prev_reqs):
+            self._traverse_apply(
+                r.docs,
+                doc_pointers=doc_pointers,
+                *args, **kwargs
+            )
+
+        self._traverse_apply(
+            self.req.docs,
+            doc_pointers=doc_pointers,
+            concatenate_by_modality=True,
+            *args, **kwargs
+        )
+
+    def _apply_all(
+            self,
+            docs: Iterable['jina_pb2.Document'],
+            context_doc: 'jina_pb2.Document',
+            field: str,
+            doc_pointers: Dict,
+            concatenate_by_modality: bool = False,
+            *args,
+            **kwargs
+    ) -> None:
+        # docs are chunks of context_doc returned by traversal rec.
+        # Group chunks which has the same modality
+        modal_doc = {}
+        for doc in docs:
+            modal = doc.modality
+            if doc.id in modal_doc:
+                modal_doc[modal].append(doc.id)
+            else:
+                modal_doc[modal] = [doc.id]
+
+        for modal, doc_ids in modal_doc.items():
```

Comment:
Perhaps someone knows better than me: In the past python dictionaries where not order preserving. So when you iterate a dictionary, you can get the items in random order. If this is true, the concatenated embedding is not deterministic. I guess that is bad, right?

Anyhow, I am not sure, whether this is still true for current python version.

Reply:
thanks for the comment, you're right, that's why we have `OrderedDict`. While concatenating is a temporary solution for this design (to be discussed further), so it's not an issue at the moment.

Reply:
Yes concatrnation is a placeholder that needs to be changed by the call to the executor